### PR TITLE
[3.3.5][6.x]Core/SmartScript: create SMART_ACTION_SET_CORPSE_DELAY

### PR DESCRIPTION
**Pull request by ghost**
by **Vincent-Michael**, updates #16488

New smart action SMART_ACTION_SET_CORPSE_DELAY:
- Allows setting the delay of corpse decay time in SmartScript

(PR opened on request from Rushor)
Constructive feedback is welcome

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 1]
```diff
@@ -2330,6 +2330,21 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
                 break;
             }
         }
+        case SMART_ACTION_SET_CORPSE_DELAY:
+        {
+            ObjectList* targets = GetTargets(e, unit);
+            if (!targets)
+                break;
+
+            for (ObjectList::const_iterator itr = targets->begin(); itr != targets->end(); ++itr)
```
Comment (MitchesD) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: why not range-based for?

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 2]
```diff
@@ -551,8 +551,9 @@ enum SMART_ACTION
     SMART_ACTION_START_CLOSEST_WAYPOINT             = 113,    // wp1, wp2, wp3, wp4, wp5, wp6, wp7
     SMART_ACTION_RISE_UP                            = 114,    // distance
     SMART_ACTION_RANDOM_SOUND                       = 115,    // soundId1, soundId2, soundId3, soundId4, soundId5, onlySelf
+    SMART_ACTION_SET_CORPSE_DELAY                   = 116,
```
Comment (DDuarte) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: The PR looks good. Add a comment to this line (`// timer`) I guess

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 1 above)
Comment (tkrokli) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: Need answer from @Rushor or @Vincent-Michael for a proper explanation.

Comment (Rushor) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: basically it's a copy from SMART_ACTION_FORCE_DESPAWN, going to test it now

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h (hunk 2 above)
Comment (tkrokli) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: Sure, anything you want inserted, I will add it. Looks like the other comments are referencing where it is used or which DB columns are linked / affected by the action.

**Comment by Rushor** [comment]
kay tested ingame with following simple script

```sql
-- Astranaar Sentinel SAI
SET @ENTRY := 6087;
UPDATE `creature_template` SET `AIName`="SmartAI" WHERE `entry`=@ENTRY;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,0,0,0,25,0,100,0,0,0,0,0,116,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Astranaar Sentinel - On Reset - corpse delay timer 1 sec");
UPDATE `creature` SET `spawntimesecs`=1 WHERE  `id`=6087;
```

so basically the creature should no behave like:
kill > 1 seconds despawn > respawn

and that's how it works with this pr.

ps: all credits should go to vincent btw, he's the autor of all :D

**Comment by tkrokli** [comment]
Thx for info, will update the commit text as well as the initial comment text in this PR. :)

File: src/server/game/AI/SmartScripts/SmartScript.cpp [hunk 3]
```diff
@@ -2330,6 +2330,21 @@ void SmartScript::ProcessAction(SmartScriptHolder& e, Unit* unit, uint32 var0, u
...
+            {
+                if (IsCreature(*itr))
```
Comment (jackpoz) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: this kind of check usually stores the result of ToCreature() to a variable in the if ( ) check and then uses it, unless I'm missing something particular about IsCreature() compared to ToCreature()

Comment (Shauren) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: no, just your regular sai code being bad - https://github.com/tkrokli/TrinityCore/blob/smart_action_set_corpse_delay/src/server/game/AI/SmartScripts/SmartScript.h#L76

Comment (jackpoz) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: @MitchesD what about this comment ?

**Comment by Rushor** [comment]
thx :)

File: src/server/game/AI/SmartScripts/SmartScript.cpp (hunk 3 above)
Comment (MitchesD) [inline src/server/game/AI/SmartScripts/SmartScript.cpp]: @jackpoz what you want to hear, that's SAI shit, check the whole file, it's used on so many places, it needs some bigger changes at all

**Comment by tkrokli** [comment]
You are welcome. I am glad I could be of help. :+1:

**Pull request merged**